Issue 11124 fixes/enhances Hazelcast implementation

agomez-dotcms opened the pull request.

agomez-dotcms committed:
#11124: Fixed hazelcast util shutdown

agomez-dotcms committed:
#11124: Made manual-cluster config wire with hazel

agomez-dotcms committed:
#11124: Minor clean-up with jgroups-cache-transport

agomez-dotcms committed:
Merge branch 'issue-11124-hookup-autocluster-to-hazelcast' of https://github.com/dotCMS/core.git into issue-11124-hookup-autocluster-to-hazelcast

agomez-dotcms committed:
#11124: Made network-tab work with hazel

agomez-dotcms committed:
Merge branch 'issue-11124-hookup-autocluster-to-hazelcast' of https://github.com/dotCMS/core.git into issue-11124-hookup-autocluster-to-hazelcast

agomez-dotcms committed:
#11124: Prevent initialization on shutdown because of invocation of
ClusterFactory's removeNodeFromCluster() from LicenseManager

agomez-dotcms committed:
Merge branch 'issue-11124-hookup-autocluster-to-hazelcast' of https://github.com/dotCMS/core.git into issue-11124-hookup-autocluster-to-hazelcast

agomez-dotcms committed:
#11124: Added support for multiple hazelcast instances

agomez-dotcms committed:
Merge branch 'issue-11124-hookup-autocluster-to-hazelcast' of https://github.com/dotCMS/core.git into issue-11124-hookup-autocluster-to-hazelcast

agomez-dotcms committed:
#11124: Added embedded and client hazelcast-transports

agomez-dotcms committed:
Merge branch 'issue-11124-hookup-autocluster-to-hazelcast' of https://github.com/dotCMS/core.git into issue-11124-hookup-autocluster-to-hazelcast

agomez-dotcms committed:
#11124: Fixed hazel auto-cluster w\ fresh/starter instances.
Successfully tested on both embedded and client hazel-transport against
both fresh/starter and non-starter (re-started) instances

agomez-dotcms committed:
Merge branch 'issue-11124-hookup-autocluster-to-hazelcast' of https://github.com/dotCMS/core.git into issue-11124-hookup-autocluster-to-hazelcast

agomez-dotcms committed:
#11124: Added embedded and client hazel-cache-providers

agomez-dotcms committed:
Merge branch 'issue-11124-hookup-autocluster-to-hazelcast' of https://github.com/dotCMS/core.git into issue-11124-hookup-autocluster-to-hazelcast

agomez-dotcms committed:
Merge branch 'issue-11124-hookup-autocluster-to-hazelcast' of https://github.com/dotCMS/core.git into issue-11124-hookup-autocluster-to-hazelcast

agomez-dotcms committed:
#11124: Fixed hazel-not-active exception when applying cluster-license

agomez-dotcms committed:
Merge branch 'issue-11124-hookup-autocluster-to-hazelcast' of https://github.com/dotCMS/core.git into issue-11124-hookup-autocluster-to-hazelcast

agomez-dotcms committed:
#11124: Not to send cache-invalidation-messages when a cache-chain is
totally distributed for a particular region

agomez-dotcms committed:
Merge branch 'master' of https://github.com/dotCMS/core into issue-11124-hookup-autocluster-to-hazelcast

jdotcms commented on dotCMS/src/main/java/com/dotcms/cache/transport/AbstractHazelcastCacheTransport.java:
```diff
@@ -11,35 +12,43 @@
 
     private Map<String, Map<String, Boolean>> cacheStatus;
+
+    private AtomicLong receivedMessages = new AtomicLong(0);
```
evaluate if these properties might be final

jdotcms commented on dotCMS/src/main/java/com/dotcms/cache/transport/AbstractHazelcastCacheTransport.java:
```diff
@@ -11,35 +12,43 @@
     private final String topicName = "dotCMSClusterCacheInvalidation";
     private String topicId;
 
+    private boolean isInitialized;
```
This may be a AtomicBoolean for concurrency.

jdotcms commented on dotCMS/src/main/java/com/dotcms/cluster/business/HazelcastUtil.java:
```diff
@@ -1,74 +1,198 @@
+	    EMBEDDED("hazelcast-embedded.xml"),
+	    CLIENT("hazelcast-client.xml");
+
+	    private String path;
```
this must be final

jdotcms commented on dotCMS/src/main/java/com/dotcms/cluster/business/HazelcastUtil.java:
```diff
@@ -1,74 +1,198 @@
+	    }
+	}
+
+    private static Map<HazelcastInstanceType, HazelcastInstance> _memberInstances = new HashMap<>();
```
evaluate if it is better to use ConcurrentHashMap instead of syncr blocks

jdotcms commented on dotCMS/src/main/java/com/dotcms/cluster/business/HazelcastUtil.java:
```diff
@@ -1,74 +1,198 @@
+		    return Hazelcast.newHazelcastInstance(config);
+		} finally {
+	        try {
+	            is.close();
```
you can use CloseUtils to do quiet closing.

agomez-dotcms committed:
#11414: Incorporated feedback after PR (blindly)

agomez-dotcms committed:
#11414: Incorporated feedback after PR (blindly)

agomez-dotcms committed:
#11414: Incorporated feedback after PR (blindly)

agomez-dotcms committed:
#11414: Incorporated feedback after PR (blindly)

dsilvam commented on dotCMS/src/main/java/com/dotcms/cluster/business/HazelcastUtil.java:
```diff
@@ -1,74 +1,198 @@
-                            }
-                        }
+
+                    if (instanceType == HazelcastInstanceType.EMBEDDED) {
```
@agomez-dotcms can we move this code into a Factory? HazelCastInstanceFactory, and have a getInstance(String xmlPath) method that has the if else and proper creation?

agomez-dotcms committed:
#11124: Changes in Auto-wire related properties
(renamings/defaults/usages):
1. Change default of "CLUSTER_AUTOWIRE" from "true" to "false"
2. Rename "CLUSTER_HAZELCAST_AUTOWIRE" with "AUTOWIRE_CLUSTER_TRANSPORT"
(default "true"). Usages in disjuntion with "CLUSTER_AUTOWIRE"
3. Rename "AUTOWIRE_MANAGE_ES_REPLICA" with "AUTOWIRE_CLUSTER_ES"
(default "true"). Usages in disjuntion with "CLUSTER_AUTOWIRE"

agomez-dotcms committed:
Merge branch 'issue-11124-hookup-autocluster-to-hazelcast' of https://github.com/dotCMS/core.git into issue-11124-hookup-autocluster-to-hazelcast

agomez-dotcms committed:
Merge branch 'issue-11124-hookup-autocluster-to-hazelcast' of https://github.com/dotCMS/core.git into issue-11124-hookup-autocluster-to-hazelcast

agomez-dotcms committed:
#11124: Introduced design improvement hazelcastutil (singleton, blindly)

agomez-dotcms committed:
#11124: Introduced requested performance optimization for not spreading
cache-invalidations repeatedly in distributed caches over cluster-nodes

agomez-dotcms committed:
Merge branch 'issue-11124-hookup-autocluster-to-hazelcast' of https://github.com/dotCMS/core.git into issue-11124-hookup-autocluster-to-hazelcast

agomez-dotcms committed:
#11124: Changes in Auto-wire related properties
(renamings/defaults/usages):
1. Change default of "CLUSTER_AUTOWIRE" from "true" to "false"
2. Rename "CLUSTER_HAZELCAST_AUTOWIRE" with "AUTOWIRE_CLUSTER_TRANSPORT"
(default "true"). Usages in disjuntion with "CLUSTER_AUTOWIRE"
3. Leave "AUTOWIRE_MANAGE_ES_REPLICA" (default "true"). Usages in
conjuntion with "CLUSTER_AUTOWIRE"
4. Introduce "AUTOWIRE_CLUSTER_ES" with
(default "true"). Usages in disjuntion with "CLUSTER_AUTOWIRE"

agomez-dotcms committed:
Merge branch 'issue-11124-hookup-autocluster-to-hazelcast' of https://github.com/dotCMS/core.git into issue-11124-hookup-autocluster-to-hazelcast

agomez-dotcms committed:
Merge remote-tracking branch 'origin/master' into issue-11124-hookup-autocluster-to-hazelcast

agomez-dotcms committed:
#11124/#11463: Fixed number of nodes on Hazelcast cluster

agomez-dotcms committed:
#11124 / #11468: Removed CLUSTER_AUTOWIRE property

agomez-dotcms committed:
#11124/#11465: Fixed class-not-found-exception on Hazelcast client

agomez-dotcms committed:
Merge branch 'issue-11124-hookup-autocluster-to-hazelcast' of https://github.com/dotCMS/core.git into issue-11124-hookup-autocluster-to-hazelcast

agomez-dotcms committed:
Merge branch 'issue-11124-hookup-autocluster-to-hazelcast' of https://github.com/dotCMS/core.git into issue-11124-hookup-autocluster-to-hazelcast

agomez-dotcms committed:
#11124 / #11468: Removed CLUSTER_AUTOWIRE property

agomez-dotcms committed:
#11124 / #11464: Cleaned-up cache-stats for Hazelcast client
